### PR TITLE
Add limit parameter to nafill and setnafill

### LINKED ISSUE #7677: Could a parameter maxfill be added to na.fill
Could a parameter maxfill be added to nafill? maxfill is an integer. When the number of consecutive NA values filled downward by a non-NA value exceeds maxfill, the filling stops until the next non-NA value appears. This is practical in time series processing, as overly old historical data is not suitable for continued filling.

### PR DESCRIPTION
**Pull request by venom1204**
closes #7677

this PR adds a limit argument to nafill and setnafill to restrict the maximum number of consecutive NA values filled. Default is Inf to preserve existing behavior. Updated R interface, C kernel logic, and documentation.

hi @tdhock @joshhwuu can you please have a look at this when you got time thanks .

### CODE REVIEW

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -21831,3 +21831,9 @@ test(2378.92, .Call(CresizeVector, x, 2:3), error = "must be length 1 non-NA non
 test(2378.93, .Call(CresizeVector, x, -2L), error = "must be length 1 non-NA non-negative integer value")
 test(2378.94, .Call(CresizeVector, x, NA_integer_), error = "must be length 1 non-NA non-negative integer value")
 test(2378.95, .Call(CresizeVector, NULL, 2L), error = "must be a vector")
+
+# #7677: add limit argument to nafill()
+test(2379.01, nafill(c(1, NA, NA, NA, 5), type="locf", limit=1), c(1, 1, NA, NA, 5))
+test(2379.02, nafill(c(1, NA, NA, 5, NA, NA, NA, 9), type="locf", limit=2), c(1, 1, 1, 5, 5, 5, NA, 9))
+test(2379.03, nafill(c(1, NA, NA, 4, NA), type="locf", limit=Inf), c(1, 1, 1, 4, 4))
+test(2379.04, nafill(c(NA, NA, 3, NA, NA), type="locf", limit=1), c(NA, NA, 3, 3, NA))
```
Comment (ben-schwen) [inline inst/tests/tests.Rraw]: whats the difference between 01 and 04?

Comment (ben-schwen) [reply inst/tests/tests.Rraw]: can we also test edge cases like negative, 0, NA, NaN, NULL?

File: man/nafill.Rd [hunk 2]
```diff
@@ -51,6 +52,12 @@ nafill(dt, "nocb")
 
 setnafill(dt, "locf", cols=c("v2","v3"))
 dt
+
+# limit= restricts the number of consecutive fills
+y = c(1, NA, NA, NA, 5)
+nafill(y, "locf", limit=1) # Only fills the first NA
+nafill(y, "locf", limit=2) # Fills the first two NAs
```
Comment (ben-schwen) [inline man/nafill.Rd]: I dont get the added value of the second line, maybe change it to `Inf` for showcasing filling all.

File: R/shift.R [hunk 3]
```diff
@@ -26,12 +26,16 @@ shift = function(x, n=1L, fill, type=c("lag", "lead", "shift", "cyclic"), give.n
   ans
 }
 
-nafill = function(x, type=c("const","locf","nocb"), fill=NA, nan=NA) {
+nafill = function(x, type=c("const","locf","nocb"), fill=NA, nan=NA, limit=Inf) {
   type = match.arg(type)
-  .Call(CnafillR, x, type, fill, nan_is_na(nan), FALSE, NULL)
+  if (!is.numeric(limit) || length(limit) != 1L || limit < 0)
+     stopf("limit must be a non-negative scalar numeric or Inf")
```
Comment (ben-schwen) [inline R/shift.R]: unprecise since Inf is also non-negative

**Review by ben-schwen — CHANGES_REQUESTED** [review]
What about type="const". What about fractions like `limit=1.5`?

Please add tests for the following cases:
* nocb
* limit: 0, NA, NaN, negative, fractional, complex, NULL, and length > 1
* integer and character
* setnafill() by-reference

NEWS is missing.

What `nafill(c(NA, 1), "locf", fill=0, limit=0)`

File: src/nafill.c [hunk 4]
```diff
@@ -15,34 +15,52 @@ void nafillDouble(double *x, uint_fast64_t nx, unsigned int type, double fill, b
       }
     }
   } else if (type==1) { // locf
+    uint_fast64_t fills = 0;
     if (nan_is_na) {
       ans->dbl_v[0] = ISNAN(x[0]) ? fill : x[0];
+      if (ISNAN(x[0])) fills = 1;
```
Comment (ben-schwen) [inline src/nafill.c]: We do not need this double branching on `ISNAN(x[0])`. This will also happen at the other cases below!

File: src/nafill.c [hunk 5]
```diff
@@ -128,6 +176,8 @@ SEXP nafillR(SEXP obj, SEXP type, SEXP fill, SEXP nan_is_na_arg, SEXP inplace, S
   if (verbose)
     tic = omp_get_wtime();
 
+  double limit_val = REAL(limit)[0];
```
Comment (ben-schwen) [inline src/nafill.c]: Might be cleaner to represent as `uint64_fast64_t` since we also compare with it later

```c
const double limit_d = REAL(limit)[0];
const uint_fast64_t limit_n = !R_FINITE(limit_d) || limit_d >= (double)UINT_FAST64_MAX ? UINT_FAST64_MAX : (uint_fast64_t)limit_d;
```

**Review by ben-schwen — CHANGES_REQUESTED** [review]
Ty for the changes. We are almost there:
Still missing.

* the double branching thingy
* the c conversion of limit (this I consider as optional but it seems the right/cleaner approach)
* clarify in the docs whats happening for fractions of limit
* clarify whats happening for `type='const'`
* clarify boundary case for `nafill(c(NA, 1), "locf", fill=0, limit=0)` which gives `c(0, 1)`
* code coverage is still failing

File: src/nafill.c [hunk 6]
```diff
@@ -1,48 +1,43 @@
 #include "data.table.h"
 
-void nafillDouble(double *x, uint_fast64_t nx, unsigned int type, double fill, bool nan_is_na, ans_t *ans, bool verbose) {
+void nafillDouble(double *x, uint_fast64_t nx, unsigned int type, double fill, bool nan_is_na, ans_t *ans, bool verbose, uint_fast64_t limit) {
   double tic=0.0;
   if (verbose)
     tic = omp_get_wtime();
   if (type==0) { // const
-    if (nan_is_na) {
```
Comment (ben-schwen) [inline src/nafill.c:8]: why do we remove this explicit branching? please dont just revert but explain!

File: src/nafill.c [hunk 7]
```diff
@@ -210,15 +224,7 @@ SEXP nafillR(SEXP obj, SEXP type, SEXP fill, SEXP nan_is_na_arg, SEXP inplace, S
     }
   }
 
-  unsigned int itype=-1;
```
Comment (ben-schwen) [inline src/nafill.c:213]: why do we make this change here? please explain!

File: inst/tests/tests.Rraw [hunk 8]
```diff
@@ -21952,3 +21952,27 @@ test(2385.04, setorderv(DT, order=c(x=1L, x=-1L)), error="order argument has nam
...
+test(2380.02, nafill(c(1, NA, NA, 5, NA, NA, NA, 9), type="locf", limit=2), c(1, 1, 1, 5, 5, 5, NA, 9))
+test(2380.03, nafill(c(1, NA, NA, 4, NA), type="locf", limit=Inf), c(1, 1, 1, 4, 4))
+test(2380.04, nafill(c(NA, NA, 3, NA, NA), type="locf", limit=1), c(NA, NA, 3, 3, NA))
+test(2380.05, nafill(c(1, NA, NA, 5), type="nocb", limit=1), c(1, NA, 5, 5))
+test(2380.06, nafill(c(1L, NA, NA), type="locf", limit=1), c(1L, 1L, NA))
+test(2380.07, nafill(c("a", NA, NA), type="locf", limit=1), c("a", "a", NA))
+test(2380.08, nafill(c(1, NA, NA), type="locf", limit=0), c(1, NA, NA))
+dt = data.table(a=c(1, NA, NA))
+setnafill(dt, type="locf", limit=1)
+test(2380.09, {dt=data.table(a=c(1,NA,NA)); setnafill(dt, type="locf", limit=1); dt$a}, c(1, 1, NA))
```
Comment (ben-schwen) [inline inst/tests/tests.Rraw]: why no test at the 2nd line?

Does also look quite more complicated than simple writing...
```suggestion
test(2380.09, setnafill(data.table(a=c(1, NA, NA)), type="locf", limit=1), data.table(a=c(1, 1, NA)))
```

File: inst/tests/tests.Rraw [hunk 9]
```diff
@@ -21952,3 +21952,27 @@ test(2385.04, setorderv(DT, order=c(x=1L, x=-1L)), error="order argument has nam
...
+test(2380.10, nafill(c(NA, 1), "locf", fill=0, limit=0), c(0, 1))
+test(2380.11, nafill(c(1, NA, NA), type="locf", limit=1.5), c(1, 1, NA))
+test(2380.12, nafill(c(1, NA, NA), type="const", fill=0, limit=1), c(1, 0, 0))
+test(2380.13, nafill(1:5, limit=-1), error="limit must be a non-negative scalar numeric")
+test(2380.14, nafill(1:5, limit=NA), error="limit must be a non-negative scalar numeric")
+test(2380.15, nafill(1:5, limit=NULL), error="limit must be a non-negative scalar numeric")
+test(2380.16, nafill(1:5, limit=c(1, 2)), error="limit must be a non-negative scalar numeric")
+test(2380.17, nafill(1:5, limit=1+2i), error="limit must be a non-negative scalar numeric")
+test(2380.18, setnafill(data.table(a=1:5), limit=-1), error="limit must be a non-negative scalar numeric")
```
Comment (ben-schwen) [inline inst/tests/tests.Rraw]: Maybe loop over?
```suggestion
lims = list(-1, NA, NaN, NULL, c(1, 2), 1+2i) 
for (i in seq_along(lims)) {
	test(2380.12 + i / 100, nafill(1:5, limit=lims[[i]]), error="limit must be a non-negative scalar numeric")
}
```

File: inst/tests/tests.Rraw [hunk 10]
```diff
@@ -21952,3 +21952,27 @@ test(2385.04, setorderv(DT, order=c(x=1L, x=-1L)), error="order argument has nam
...
 DT = data.table(a=c(2,1,2), b=3:1)
 test(2385.06, setorderv(copy(DT), order=c(a=1L)), setorderv(copy(DT), cols="a", order=1L))
+
+# #7677: add limit argument to nafill()
+test(2380.01, nafill(c(1, NA, NA, NA, 5), type="locf", limit=1), c(1, 1, NA, NA, 5))
+test(2380.02, nafill(c(1, NA, NA, 5, NA, NA, NA, 9), type="locf", limit=2), c(1, 1, 1, 5, 5, 5, NA, 9))
+test(2380.03, nafill(c(1, NA, NA, 4, NA), type="locf", limit=Inf), c(1, 1, 1, 4, 4))
+test(2380.04, nafill(c(NA, NA, 3, NA, NA), type="locf", limit=1), c(NA, NA, 3, 3, NA))
+test(2380.05, nafill(c(1, NA, NA, 5), type="nocb", limit=1), c(1, NA, 5, 5))
+test(2380.06, nafill(c(1L, NA, NA), type="locf", limit=1), c(1L, 1L, NA))
```
Comment (ben-schwen) [inline inst/tests/tests.Rraw]: test 1,2,3,6,7, could probably be subsumed in 2 tests.

File: src/nafill.c (hunk 7 above)
Comment (venom1204) [reply src/nafill.c:213]: I initially changed this to use a more compact format thinking it would be cleaner, but I'll revert it now to maintain consistency with the original style

File: src/nafill.c (hunk 6 above)
Comment (venom1204) [reply src/nafill.c:8]: I'll revert this. I thought a single loop would be cleaner

File: src/nafill.c (hunk 7 above)
Comment (ben-schwen) [reply src/nafill.c:213]: I dont get why a nested ternary operator should be ever cleaner :D

File: src/nafill.c (hunk 6 above)
Comment (ben-schwen) [reply src/nafill.c:8]: You're right that a single loop looks cleaner, but its not about what we are here after.

In the single loop you have to do the check every iteration, why otherwise you do it once. Now you could hope that the compiler is smart enough enough to optimize this after all, but you cannot be sure that it does.

This also applies to the if (nan_is_na) checks which you rewrote to single loops. I did not flag the other occurrences...

**Comment by ben-schwen** [comment]
TY!
